### PR TITLE
Add query to ensure we don't add duplicate message ids in mongo

### PR DESCRIPTION
**Pull request by olakunlekuye2**

### CODE REVIEW

File: app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala [hunk 1]
```diff
@@ -548,4 +543,46 @@ class MovementsRepositoryImpl @Inject() (
...
+  }
+
+  def verifyUniqueMessageIds(messageIds: Seq[MessageId]): EitherT[Future, MongoError, Unit] = {
+    val uniqueMessageIds = messageIds.distinct
+    if (uniqueMessageIds.isEmpty) EitherT.rightT[Future, MongoError](())
+    else {
+      val uniqueMessageIdFutures = uniqueMessageIds.map(
+        messageId => isMessageIdUnique(messageId).value
+      )
+      val uniqueMessageIdResults = Future.sequence(uniqueMessageIdFutures)
```
Comment (drnaylor-oc) [inline app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala:577]: No, don't make a query per ID. Instead, collect all the IDs and use [$in](https://www.mongodb.com/docs/manual/reference/operator/query/in/) to select in one query.

**Review by drnaylor-oc — COMMENTED** [review]
I'm gonna potentially scorch this PR but not really for the code reasons. I don't think this is needed and by reading this PR, there's enough here that convinces me that we really shouldn't be doing anything.

It is the responsibility of our code to ensure we're generating random IDs. We do that by using a clock and a random function -- the first half of an ID is based on the system time, the second bit is random. That alone should prevent duplicates from happening. 

Looking at the queries here, we're querying a lot of records, potentially. CTC in phase 4 has 1.5 million movements, so potentially five million messages. That means we have to scan 5 million message IDs which we might not be able to index due to it being in a sub document. That means a full table scan every time.

Now, message IDs need to only be unique to each movement, because we only identify a message by the combination of its movement ID and message ID. So that should reduce the burden -- and so looking at what you're doing... can't we just ensure that our in-memory object doesn't have any duplicate IDs -- and not even _go_ to Mongo to check this?

I think this ticket came about because someone accidentally inserted duplicate IDs in the code -- so this is more a protection for us than anything. If that's what we're doing, then really we should just be doing an in-memory check, nothing more.

(Sorry, I didn't really get myself involved in this ticket but now I know what it is, I'm really not sure this should go in...)

File: app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala [hunk 2]
```diff
@@ -548,4 +543,46 @@ class MovementsRepositoryImpl @Inject() (
         Future.successful(Left(UnexpectedError(Some(ex))))
     })
   }
+
+  def isMessageIdUnique(messageId: MessageId): EitherT[Future, MongoError, Unit] = {
+    val selector   = Filters.elemMatch("messages", Filters.eq("id", messageId.value))
```
[the remark this answers is not in the document]
Comment (olakunlekuye2) [reply app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala:548]: It was to ensure message IDs are unique globally and I'll go with your view

[the remark this answers is not in the document]
Comment (olakunlekuye2) [reply app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala:548]: Done

File: app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala (hunk 1 above)
Comment (olakunlekuye2) [reply app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala]: Done

File: app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala [hunk 3]
```diff
@@ -176,19 +169,21 @@ class MovementsRepositoryImpl @Inject() (
...
-            } else {
-              Left(InsertNotAcknowledged(s"Insert failed for movement $movement"))
-            }
-        }
-      case Failure(NonFatal(ex)) =>
-        Future.successful(Left(UnexpectedError(Some(ex))))
-    })
+    for {
+      _ <- {
+        if (movement.messages.isEmpty) EitherT.rightT[Future, MongoError](())
```
[the remark this answers is not in the document]
Comment (olakunlekuye2) [reply app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala]: Done

File: app/uk/gov/hmrc/transitmovements/controllers/MovementsController.scala [hunk 4]
```diff
@@ -91,61 +91,66 @@ class MovementsController @Inject() (
...
+          .create(movementId, MessageType.ArrivalNotification, arrivalData.generationDate, received, None, size, request.body, MessageStatus.Processing)
+          .asPresentation
+        movement = movementFactory.createArrival(movementId, eori, MovementType.Arrival, arrivalData, message, received, received)
+        _ <- repo.insert(movement).asPresentation
+      } yield MovementResponse(movement._id, Some(movement.messages.head.id))
+
+      result.value.map {
+        case Left(baseError) => Status(baseError.code.statusCode)(Json.toJson(baseError))
+        case Right(response) => Ok(Json.toJson(response))
+      }
```
[the remark this answers is not in the document]
Comment (olakunlekuye2) [reply app/uk/gov/hmrc/transitmovements/controllers/MovementsController.scala]: Done

**Comment by olakunlekuye2** [comment]
This is no longer required

**Pull request closed without merging**